src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -266,6 +267,59 @@ public IEnumerable<ITagSpan<IClassificationTag>> GetTags(NormalizedSnapshotSpanC
                 }
             }
 
+            /// <summary>
+            /// Force computation of up to date tags.
+            /// This is called by the editor for scenarios such as print or rich text copy with classifications enabled.
+            /// Note that the editor will block the UI until results are returned.
+            /// </summary>
+            public IEnumerable<ITagSpan<IClassificationTag>> GetAllTags(NormalizedSnapshotSpanCollection spans, CancellationToken cancellationToken)
+            {
+                // Retrieve the last snapshot we computed.
+                // Needs to be done under a lock as these can be changed on the background thread.
+                ITextSnapshot lastSnapshot;
+                Document lastDocument;
+
+                lock (_gate)
+                {
+                    lastSnapshot = _lastProcessedSnapshot;
+                    lastDocument = _lastProcessedDocument;
+                }
+
+                if (spans.IsEmpty() || _workspace == null)
+                {
+                    return SpecializedCollections.EmptyEnumerable<ITagSpan<IClassificationTag>>();
+                }
+
+                var snapshot = spans[0].Snapshot;
+                var classificationService = _workspace.Services.GetLanguageServices(snapshot.ContentType)?.GetService<IClassificationService>();
```

Comment:
1. this feels weird.  why do we do this, instead of just using the `Document lastDocument;`?